src/Nethermind/Nethermind.Network.Discovery/Serializers/DiscoveryMsgSerializerBase.cs

```diff
@@ -26,68 +30,134 @@ public abstract class DiscoveryMsgSerializerBase
         }
+        IByteBuffer data = msg.Slice(98, msg.ReadableBytes - 98);
+        Memory<byte> msgBytes = msg.ReadAllBytesAsMemory();
+        Memory<byte> mdc = msgBytes.Slice(0, 32);
+        Span<byte> sigAndData = msgBytes.Span.Slice(32);
+        Span<byte> computedMdc = ValueKeccak.Compute(sigAndData).BytesAsSpan;
 
-        byte[] mdc = msg.Slice(0, 32);
-        Span<byte> signature = msg.AsSpan(32, 65);
-        // var type = new[] { msg[97] };
-        byte[] data = msg.Slice(98, msg.Length - 98);
-        Span<byte> computedMdc = ValueKeccak.Compute(msg.AsSpan(32)).BytesAsSpan;
-
-        if (!Bytes.AreEqual(mdc, computedMdc))
+        if (!Bytes.AreEqual(mdc.Span, computedMdc))
         {
             throw new NetworkingException("Invalid MDC", NetworkExceptionType.Validation);
         }
 
-        PublicKey nodeId = _nodeIdResolver.GetNodeId(signature.Slice(0, 64).ToArray(), signature[64], msg.AsSpan(97, msg.Length - 97));
+        PublicKey nodeId = _nodeIdResolver.GetNodeId(sigAndData.Slice(0, 64), sigAndData[64], sigAndData.Slice(65, sigAndData.Length - 65));
         return (nodeId, mdc, data);
     }
 
-    protected static Rlp Encode(IPEndPoint address)
+    protected static void Encode(RlpStream stream, IPEndPoint address, int length)
+    {
+        stream.StartSequence(length);
+        stream.Encode(address.Address.GetAddressBytes());
+        //tcp port
+        stream.Encode(address.Port);
+        //udp port
+        stream.Encode(address.Port);
+    }
+
+    protected static int GetIPEndPointLength(IPEndPoint address)
+    {
+        int length = Rlp.LengthOf(address.Address.GetAddressBytes());
+        length += Rlp.LengthOf(address.Port);
+        length += Rlp.LengthOf(address.Port);
+        return length;
+    }
+
+    protected static void SerializeNode(RlpStream stream, IPEndPoint address, byte[] id)
+    {
+        int length = GetLengthSerializeNode(address, id);
+        stream.StartSequence(length);
+        stream.Encode(address.Address.GetAddressBytes());
+        //tcp port
+        stream.Encode(address.Port);
+        //udp port
+        stream.Encode(address.Port);
+        stream.Encode(id);
+    }
+
+    protected static int GetLengthSerializeNode(IPEndPoint address, byte[] id)
     {
-        return Rlp.Encode(
-            Rlp.Encode(address.Address.GetAddressBytes()),
-            //tcp port
-            Rlp.Encode(address.Port),
-            //udp port
-            Rlp.Encode(address.Port)
-        );
+        int length = Rlp.LengthOf(address.Address.GetAddressBytes());
+        length += Rlp.LengthOf(address.Port);
+        length += Rlp.LengthOf(address.Port);
+        length += Rlp.LengthOf(id);
+        return length;
     }
 
-    protected static Rlp SerializeNode(IPEndPoint address, byte[] id)
+    protected void PrepareBufferForSerialization(IByteBuffer byteBuffer, int dataLength, byte msgType)
```

Comment:
You can also do something like this so that people wont forget to sign it.

```
    protected void SignEncodedMessage<T>(IByteBuffer byteBuffer, int contentLength, byte msgType, T msg, Action<T, int, IByteBuffer> encoder)
    {
        int dataLength = Rlp.LengthOfSequence(contentLength);

        byteBuffer.MarkIndex();
        byteBuffer.EnsureWritable(MdcSigOffset + 1 + dataLength);
        byteBuffer.SetWriterIndex(byteBuffer.WriterIndex + MdcSigOffset);
        byteBuffer.WriteByte(msgType);

        encoder.Invoke(msg, contentLength, byteBuffer);

        byteBuffer.ResetIndex();
        AddSignatureAndMdc(byteBuffer, dataLength + 1);
    }
```

Use like:

```
    public void Serialize(IByteBuffer byteBuffer, EnrResponseMsg msg)
    {
        int contentLength = Rlp.LengthOfKeccakRlp;
        contentLength += msg.NodeRecord.GetRlpLengthWithSignature();

        SignEncodedMessage(byteBuffer, contentLength, (byte)msg.MsgType, msg, static (msg, contentLength, buffer) =>
        {
            NettyRlpStream rlpStream = new(buffer);
            rlpStream.StartSequence(contentLength);
            rlpStream.Encode(msg.RequestKeccak);
            msg.NodeRecord.Encode(rlpStream);
        });
    }
```

Reply:
for messages where `int dataLength = Rlp.LengthOfSequence(contentLength);` is not valid - this would be a problem.
eg. NeighborsMsgSerializer